libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -174,12 +167,40 @@ private static string GetBlobName(string key)
                         AccessCondition.GenerateIfMatchCondition(calculatedETag),
                         new BlobRequestOptions(),
                         new OperationContext()))
-                    using (var streamWriter = new StreamWriter(blobStream))
-                    using (var jsonWriter = new JsonTextWriter(streamWriter))
+                    using (var jsonWriter = new JsonTextWriter(new StreamWriter(blobStream)))
                     {
                         JsonSerializer.Serialize(jsonWriter, newValue);
                     }
                 }));
         }
+
+        private static string GetBlobName(string key)
+        {
+            if (string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+
+            var blobName = HttpUtility.UrlEncode(key);
+            NameValidator.ValidateBlobName(blobName);
+            return blobName;
+        }
+
+        private ValueTask<CloudBlobContainer> GetBlobContainer()
+        {
+            if(_container != null)
+            {
+                return new ValueTask<CloudBlobContainer>(_container);

```

Comment:
>new ValueTask<CloudBlobContainer>( [](start = 23, length = 34)

wait a minute, the docs I read said you should be able to return the _container itself here.  Did I read the wrong docs?

Reply:
Yeah, you'll notice I didn't use the `async` keyword on the method, so I'm in charge of having to construct the `ValueTask<T>` manually myself in that case. Here's the difference in IL:

Without `async` keyword:

```MSIL
.method private hidebysig instance valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> GetBlobContainer () cil managed 
    {
        .locals init (
            [0] bool V_0,
            [1] valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> V_1
        )

        IL_0000: nop
        IL_0001: ldarg.0
        IL_0002: ldfld class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer Microsoft.Bot.Builder.Azure.AzureBlobStorage::_container
        IL_0007: ldnull
        IL_0008: cgt.un
        IL_000a: stloc.0
        IL_000b: ldloc.0
        IL_000c: brfalse.s IL_001d

        IL_000e: nop
        IL_000f: ldarg.0
        IL_0010: ldfld class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer Microsoft.Bot.Builder.Azure.AzureBlobStorage::_container
        IL_0015: newobj instance void valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer>::.ctor(!0)
        IL_001a: stloc.1
        IL_001b: br.s IL_002b

        IL_001d: ldarg.0
        IL_001e: call instance class [netstandard]System.Threading.Tasks.Task`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> Microsoft.Bot.Builder.Azure.AzureBlobStorage::'<GetBlobContainer>g__EnsureBlobContainerExists|10_0'()
        IL_0023: newobj instance void valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer>::.ctor(class [netstandard]System.Threading.Tasks.Task`1<!0>)
        IL_0028: stloc.1
        IL_0029: br.s IL_002b

        IL_002b: ldloc.1
        IL_002c: ret
    }
```

With `async` keyword:

```MSIL
.method private hidebysig instance valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> GetBlobContainer () cil managed 
    {
        .custom instance void [netstandard]System.Runtime.CompilerServices.AsyncStateMachineAttribute::.ctor(class [netstandard]System.Type) = (
            01 00 44 4d 69 63 72 6f 73 6f 66 74 2e 42 6f 74
            2e 42 75 69 6c 64 65 72 2e 41 7a 75 72 65 2e 41
            7a 75 72 65 42 6c 6f 62 53 74 6f 72 61 67 65 2b
            3c 47 65 74 42 6c 6f 62 43 6f 6e 74 61 69 6e 65
            72 3e 64 5f 5f 31 30 00 00
        )
        .custom instance void [netstandard]System.Diagnostics.DebuggerStepThroughAttribute::.ctor() = (
            01 00 00 00
        )
        .locals init (
            [0] class Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10' V_0,
            [1] valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> V_1
        )

        IL_0000: newobj instance void Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::.ctor()
        IL_0005: stloc.0
        IL_0006: ldloc.0
        IL_0007: ldarg.0
        IL_0008: stfld class Microsoft.Bot.Builder.Azure.AzureBlobStorage Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::'<>4__this'
        IL_000d: ldloc.0
        IL_000e: call valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer>::Create()
        IL_0013: stfld valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::'<>t__builder'
        IL_0018: ldloc.0
        IL_0019: ldc.i4.m1
        IL_001a: stfld int32 Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::'<>1__state'
        IL_001f: ldloc.0
        IL_0020: ldfld valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::'<>t__builder'
        IL_0025: stloc.1
        IL_0026: ldloca.s V_1
        IL_0028: ldloca.s V_0
        IL_002a: call instance void valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer>::Start<class Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'>(!!0&)
        IL_002f: ldloc.0
        IL_0030: ldflda valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::'<>t__builder'
        IL_0035: call instance valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer>::get_Task()
        IL_003a: ret
    }
```

So while the `async` one might read a little better, it's always costing you an allocation of that async state machine which I think is worthwhile optimizing away in a case like this with a little more explicit construction of the `ValueTask<T>`.

Reply:
@tomlm, take that! :) 

heh.